platform-api/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/WsAgentLauncherImpl.java

```diff
@@ -0,0 +1,113 @@
+import org.eclipse.che.api.machine.server.model.impl.CommandImpl;
+import org.eclipse.che.api.machine.server.spi.Instance;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.ws.rs.HttpMethod;
+import javax.ws.rs.core.UriBuilder;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.URI;
+
+/**
+ * Starts ws agent in the machine and wait until ws agent sends notification about its start
+ *
+ * @author Alexander Garagatyi
+ */
+@Singleton
+public class WsAgentLauncherImpl implements WsAgentLauncher {
+    public static final String WS_AGENT_PROCESS_START_COMMAND  = "machine.server.ext.run_command";
+    public static final String WS_AGENT_PROCESS_OUTPUT_CHANNEL = "workspace:%s:ext-server:output";
+    public static final String WS_AGENT_PROCESS_NAME           = "CheWsAgent";
+    public static final int    WS_AGENT_PORT                   = 4401;
+
+    private final MachineManager         machineManager;
+    private final HttpJsonRequestFactory httpJsonRequestFactory;
+    private final String                 wsAgentStartCommandLine;
+    private final long                   wsAgentMaxStartTimeMs;
+    private final long                   wsAgentPingDelayMs;
+    private final int                    wsAgentPingConnectionTimeoutMs;
+    private final String                 wsAgentPingPath;
+
+    @Inject
+    public WsAgentLauncherImpl(MachineManager machineManager,
+                               HttpJsonRequestFactory httpJsonRequestFactory,
+                               @Named(WS_AGENT_PROCESS_START_COMMAND) String wsAgentStartCommandLine,
+                               @Named("api.endpoint") URI apiEndpoint,
+                               @Named("machine.ws-agent.max-start-time-ms") long wsAgentMaxStartTimeMs,
+                               @Named("machine.ws-agent.ping-delay-ms") long wsAgentPingDelayMs,
+                               @Named("machine.ws-agent.ping-conn-timeout-ms") int wsAgentPingConnectionTimeoutMs) {
+        this.machineManager = machineManager;
+        this.httpJsonRequestFactory = httpJsonRequestFactory;
+        this.wsAgentStartCommandLine = wsAgentStartCommandLine;
+        this.wsAgentMaxStartTimeMs = wsAgentMaxStartTimeMs;
+        this.wsAgentPingDelayMs = wsAgentPingDelayMs;
+        this.wsAgentPingConnectionTimeoutMs = wsAgentPingConnectionTimeoutMs;
+        this.wsAgentPingPath = apiEndpoint.getPath();
+    }
+
+    @Override
+    public void startWsAgent(String workspaceId) throws NotFoundException, MachineException, InterruptedException {
+        final Instance devMachine = machineManager.getDevMachine(workspaceId);
+        try {
+            machineManager.exec(devMachine.getId(),
+                                new CommandImpl(WS_AGENT_PROCESS_NAME, wsAgentStartCommandLine, "Arbitrary"),
+                                String.format(WS_AGENT_PROCESS_OUTPUT_CHANNEL, workspaceId));
```

Comment:
I think in this case it is better to have `public static` method for getting channel.
Why:
The modifier of the `WS_AGENT_PROCESS_OUTPUT_CHANNEL` is public, which means that you want to reuse it in the other place, so you will do the same operation there `String.format(WS_AGENT_PROCESS_OUTPUT_CHANNEL, workspaceId)`.  If you create the method for it the usage will be obvious and won't let the developer to miss something